[NVIDIA] upstream FA4

johnnynunez opened the pull request:
1. Added num_splits_heuristic function (lines 39-46)
Heuristic to determine the number of splits for split KV computation
2. Added new parameters to _flash_attn_fwd
num_splits: int = 1 (line 70)
mask_mod: Optional[Callable] = None (line 74)
3. Updated window_size handling logic (lines 225-236)
Supports mask_mod parameter
Handles causal/local window logic with mask_mod
4. Added split KV logic (lines 237-260)
Computes is_split_kv based on num_splits
Creates out_partial and lse_partial tensors when split KV is enabled
Automatic split calculation when num_splits < 1
5. Updated compile key (lines 350-376)
Added score_mod_hash and mask_mod_hash
Added is_split_kv flag
Added paged_kv_non_tma flag
Added buffer count instead of aux_tensors count
6. Updated kernel initialization calls
FlashAttentionForwardSm90 (lines 382-401): Added mask_mod, intra_wg_overlap=True, mma_pv_is_rs=True
FlashAttentionForwardSm100 (lines 407-428): Added mask_mod, is_split_kv, paged_kv_non_tma, is_varlen_q, m_block_size, n_block_size
7. Added _flash_attn_fwd_combine function (lines 486-616)
Combines partial outputs from split KV computation
Includes caching logic
8. Added combine call (lines 473-481)
Calls _flash_attn_fwd_combine when is_split_kv is True
9. Added import (line 25)
from flash_attn_origin.cute.flash_fwd_combine import FlashAttentionForwardCombine

johnnynunez committed:
feat: add split attention support and combine kernel for improved efficiency

johnnynunez committed:
feat: add split attention support and combine kernel for improved efficiency

Qiaolin-Yu committed:
fix

Qiaolin-Yu committed:
Merge branch 'main' into flash-attention

johnnynunez committed:
Merge branch 'sgl-project:main' into flash-attention

johnnynunez committed:
Update flash-attention repository tag

johnnynunez committed:
Merge branch 'sgl-project:main' into flash-attention

Qiaolin-Yu committed:
upd

johnnynunez committed:
enable jetson agx thor FA4

johnnynunez committed:
Merge branch 'main' into flash-attention

Qiaolin-Yu committed:
upd

Qiaolin-Yu committed:
fix

Qiaolin-Yu committed:
fix

Qiaolin-Yu committed:
fix

HydraQYH reviewed (commented):
Hello Johnny. I'm working with @BBuf to update the flash-attention in sgl-kernel. I pulled your commit and built sgl-kernel, and I found that the flash-attention in sgl-kernel currently has a significant performance gap compared to upstream. It looks like branch `slg-kernel` in project [sgl-flash-attn](https://github.com/sgl-project/sgl-flash-attn) might need to rebase upstream. Can you help?

johnnynunez committed:
Merge branch 'sgl-project:main' into flash-attention

Qiaolin-Yu committed:
upd

Qiaolin-Yu committed:
fix

Qiaolin-Yu committed:
upd

Qiaolin-Yu committed:
fix

Qiaolin-Yu committed:
Merge remote-tracking branch 'upstream/main' into pr-15182

Fridge003 committed:
Merge branch 'main' into flash-attention

Fridge003 committed:
fix import

Fridge003 committed:
Merge branch 'main' into flash-attention

Qiaolin-Yu committed:
upd

Qiaolin-Yu commented:
> > b200 tests passed with newest cutedsl and tvm-ffi https://github.com/sgl-project/sglang/actions/runs/20890229706/job/60022405517?pr=15182
> 
> The last commit from FA4, changed a little bit the interface.py. We can update it, i think so

I have updated it.

Fridge003 committed:
little revert

johnnynunez commented:
> > > b200 tests passed with newest cutedsl and tvm-ffi https://github.com/sgl-project/sglang/actions/runs/20890229706/job/60022405517?pr=15182
> > 
> > 
> > The last commit from FA4, changed a little bit the interface.py. We can update it, i think so
> 
> I have updated it.

thanks, that fix some bugs in sm100

Fridge003 committed:
Update python/pyproject.toml
